exchanges/gateio/gateio_websocket_request_spot.go

```diff
@@ -0,0 +1,219 @@
+package gateio
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/request"
+)
+
+var (
+	errOrdersEmpty      = errors.New("orders cannot be empty")
+	errNoOrdersToCancel = errors.New("no orders to cancel")
+	errChannelEmpty     = errors.New("channel cannot be empty")
+)
+
+// WebsocketOrderPlaceSpot places an order via the websocket connection. You can
+// send multiple orders in a single request. But only for one asset route.
+func (g *Gateio) WebsocketOrderPlaceSpot(ctx context.Context, orders []WebsocketOrder) ([]WebsocketOrderResponse, error) {
+	if len(orders) == 0 {
+		return nil, errOrdersEmpty
+	}
+
+	for i := range orders {
+		if orders[i].Text == "" {
+			// API requires Text field, or it will be rejected
+			orders[i].Text = "t-" + strconv.FormatInt(g.Counter.IncrementAndGet(), 10)
+		}
+		if orders[i].CurrencyPair == "" {
+			return nil, currency.ErrCurrencyPairEmpty
+		}
+		if orders[i].Side == "" {
+			return nil, order.ErrSideIsInvalid
+		}
+		if orders[i].Amount == "" {
+			return nil, errInvalidAmount
+		}
+		if orders[i].Type == "limit" && orders[i].Price == "" {
+			return nil, errInvalidPrice
+		}
+	}
+
+	if len(orders) == 1 {
+		var singleResponse WebsocketOrderResponse
+		return []WebsocketOrderResponse{singleResponse}, g.SendWebsocketRequest(ctx, spotPlaceOrderEPL, "spot.order_place", asset.Spot, orders[0], &singleResponse, 2)
+	}
+	var resp []WebsocketOrderResponse
+	return resp, g.SendWebsocketRequest(ctx, spotBatchOrdersEPL, "spot.order_place", asset.Spot, orders, &resp, 2)
+}
+
+// WebsocketOrderCancelSpot cancels an order via the websocket connection
+func (g *Gateio) WebsocketOrderCancelSpot(ctx context.Context, orderID string, pair currency.Pair, account string) (*WebsocketOrderResponse, error) {
+	if orderID == "" {
+		return nil, order.ErrOrderIDNotSet
+	}
+	if pair.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+
+	params := &WebsocketOrderRequest{OrderID: orderID, Pair: pair.String(), Account: account}
+
+	var resp WebsocketOrderResponse
+	return &resp, g.SendWebsocketRequest(ctx, spotCancelSingleOrderEPL, "spot.order_cancel", asset.Spot, params, &resp, 1)
+}
+
+// WebsocketOrderCancelAllByIDsSpot cancels multiple orders via the websocket
+func (g *Gateio) WebsocketOrderCancelAllByIDsSpot(ctx context.Context, o []WebsocketOrderBatchRequest) ([]WebsocketCancellAllResponse, error) {
+	if len(o) == 0 {
+		return nil, errNoOrdersToCancel
```

Comment:
MTC

Reply:
At first I was like wth is MTC 😆 - [fixed](https://github.com/thrasher-corp/gocryptotrader/pull/1598/commits/8f9e9535b27452ef7328066f9560df02036963fa)